Sources/CoreMetrics/SystemMetrics.swift

```diff
@@ -0,0 +1,337 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Dispatch
+
+#if os(Linux)
+import Glibc
+#endif
+
+extension MetricsSystem {
+    /// `bootstrapWithSystemMetrics` is an one-time configuration function which globally selects the desired metrics backend
+    /// implementation, and enables system level metrics. `bootstrapWithSystemMetrics` can be called at maximum once in any given program,
+    /// calling it more than once will lead to undefined behaviour, most likely a crash.
+    ///
+    /// - parameters:
+    ///     - factory: A factory that given an identifier produces instances of metrics handlers such as `CounterHandler`, `RecorderHandler` and `TimerHandler`.
+    ///     - config: Used to configure `SystemMetrics`.
+    public static func bootstrapWithSystemMetrics(_ factory: MetricsFactory, config: SystemMetrics.Configuration) {
+        self.bootstrap(SystemMetricsFactory(factory: factory, config: config))
+    }
+
+    internal class SystemMetricsFactory: MetricsFactory {
+        fileprivate let queue = DispatchQueue(label: "com.apple.CoreMetrics.SystemMetricsHandler", qos: .background)
+        fileprivate let timeInterval: DispatchTimeInterval
+        fileprivate let dataProvider: SystemMetrics.DataProvider
+        fileprivate let labels: SystemMetrics.Labels
+        fileprivate let timer: DispatchSourceTimer
+        internal let underlying: MetricsFactory
+
+        init(factory: MetricsFactory, config: SystemMetrics.Configuration) {
+            self.underlying = factory
+            self.timeInterval = config.interval
+            self.dataProvider = config.dataProvider
+            self.labels = config.labels
+            self.timer = DispatchSource.makeTimerSource(queue: self.queue)
+            
+            timer.setEventHandler(handler: DispatchWorkItem(block: { [weak self] in
+                guard let self = self, let metrics = self.dataProvider() else { return }
+                Gauge(label: self.labels.label(for: \.virtualMemoryBytes)).record(metrics.virtualMemoryBytes)
+                Gauge(label: self.labels.label(for: \.residentMemoryBytes)).record(metrics.residentMemoryBytes)
+                Gauge(label: self.labels.label(for: \.startTimeSeconds)).record(metrics.startTimeSeconds)
+                Gauge(label: self.labels.label(for: \.cpuSecondsTotal)).record(metrics.cpuSeconds)
+                Gauge(label: self.labels.label(for: \.maxFileDescriptors)).record(metrics.maxFileDescriptors)
+                Gauge(label: self.labels.label(for: \.openFileDescriptors)).record(metrics.openFileDescriptors)
+            }))
+            
+            timer.schedule(deadline: .now() + self.timeInterval, repeating: self.timeInterval)
+            if #available(OSX 10.12, *) {
+                timer.activate()
+            } else {
+                timer.resume()
+            }
```

Comment:
Not sure exactly what the difference between `.activate()` and `.resume()` are here, but they achieve the same result. Not sure if it's acceptable to only use `.resume()` instead of both like this.

CC @drexin

Reply:
I think `resume` does a little more in case the source ran before and was suspended, but `resume` works for sources that have never been activated. I'm not sure if there are any downsides to using it in this case.